File: ai-service/src/blueprint/parser.py
"""Blueprint Parser - JSON to Blueprint with validation."""

from typing import Dict, Any, List, Union
from datetime import datetime

from core import get_logger, ValidationError
from core.json import extract_json, JSONParseError

logger = get_logger(__name__)


class BlueprintParser:
    """Parses Blueprint (.bp) files into Package format"""
    
    def __init__(self):
        self.templates = {}
        self._id_counter = 0
    
    def parse(self, bp_content: str) -> Dict[str, Any]:
        """
        Parse Blueprint JSON string to Package dict
        
        Args:
            bp_content: JSON content string
            
        Returns:
            Package dictionary compatible with types.Package
        """
        # Reset ID counter for each parse
        self._id_counter = 0
        
        # Use optimized JSON parser with automatic extraction and repair
        try:
            bp = extract_json(bp_content, repair=True)
        except JSONParseError as e:
            logger.error("json_parse_failed", error=str(e))
            raise ValidationError(f"Invalid JSON: {e}") from e
        
        if not bp or not isinstance(bp, dict):
            logger.error("invalid_format", type=type(bp).__name__)
            raise ValidationError("Invalid Blueprint format: expected JSON object")
        
        if "app" not in bp:
            logger.error("missing_app_section")
            raise ValidationError("Invalid Blueprint: missing 'app' section")
        
        app = bp["app"]
        
        # Validate required fields
        if not app.get("id"):
            logger.error("missing_field", field="app.id")
            raise ValidationError("app.id is required")
        if not app.get("name"):
            logger.error("missing_field", field="app.name")
            raise ValidationError("app.name is required")
        
        # Default timestamp
        now = datetime.utcnow().isoformat() + "Z"
        
        # Store templates for reuse
        self.templates = bp.get("templates", {})
        
        return {
            "id": app["id"],
            "name": app["name"],
            "description": app.get("description", ""),
            "icon": app.get("icon"),
            "category": app.get("category"),
            "version": app.get("version", "1.0.0"),
            "author": app.get("author", "user"),
            "created_at": now,
            "updated_at": now,
            "services": self._expand_services(bp.get("services", [])),
            "permissions": app.get("permissions", ["STANDARD"]),
            "tags": app.get("tags", []),
            "ui_spec": self._expand_ui(bp.get("ui", {})),
            "config": bp.get("config", {})  # App-specific configuration
        }
    
    def _expand_services(self, services: List[Any]) -> List[Union[str, Dict[str, Any]]]:
        """
        Expand service definitions
        
        Supports:
        - Simple strings: ["storage"] → all tools
        - Array syntax: [{storage: [get, set]}] → specific tools
        - Wildcard: [{storage: "*"}] → all tools (explicit)
        - Object config: [{storage: {tools: [get, set], scope: app}}]
        
        Returns mixed list for backward compatibility, but preserves tool info
        """
        result = []
        
        for svc in services:
            if isinstance(svc, str):
                # Simple import - all tools
                result.append(svc)
            elif isinstance(svc, dict):
                # Object with tools/config
                for key, value in svc.items():
                    if value == "*" or value is None:
                        # Wildcard or no value - all tools
                        result.append(key)
                    elif isinstance(value, list):
                        # Explicit tool list: {storage: [get, set]}
                        result.append({
                            "service": key,
                            "tools": value
                        })
                    elif isinstance(value, dict):
                        # Full config: {storage: {tools: [...], scope: app}}
                        tools = value.get("tools", "*")
                        if tools == "*":
                            result.append(key)
                        else:
                            result.append({
                                "service": key,
                                "tools": tools,
                                "config": {k: v for k, v in value.items() if k != "tools"}
                            })
                    else:
                        # Fallback - just service name
                        result.append(key)
        
        return result
    
    def _expand_ui(self, ui: Dict[str, Any]) -> Dict[str, Any]:
        """Expand UI specification with all shortcuts"""
        if not ui:
            return {
                "type": "app",
                "title": "Untitled",
                "layout": "vertical",
                "components": []
            }
        
        return {
            "type": "app",
            "title": ui.get("title", "Untitled"),
            "layout": ui.get("layout", "vertical"),
            "lifecycle_hooks": self._expand_lifecycle(ui.get("lifecycle", {})),
            "components": self._expand_components(ui.get("components", []))
        }
    
    def _expand_lifecycle(self, lifecycle: Dict[str, Any]) -> Dict[str, Any]:
        """
        Expand lifecycle hooks
        
        Supports:
        - Single action: {on_mount: "storage.get"}
        - Multiple actions: {on_mount: ["storage.get", "ui.init"]}
        """
        if not lifecycle:
            return {}
        
        result = {}
        for hook, action in lifecycle.items():
            if isinstance(action, str):
                result[hook] = [action]
            elif isinstance(action, list):
                result[hook] = action
            else:
                result[hook] = []
        
        return result
    
    def _expand_components(self, components: List[Any]) -> List[Dict[str, Any]]:
        """Recursively expand component list"""
        result = []
        
        for comp in components:
            expanded = self._expand_component(comp)
            if expanded:
                result.append(expanded)
        
        return result
    
    def _expand_component(self, comp: Any) -> Dict[str, Any]:
        """
        Expand a single component - explicit format only for streaming
        
        Format: {type: "button", id: "save", props: {...}, on_event: {...}, children: [...]}
        
        Special cases:
        - Simple strings: "Hello" -> {type: "text", id: "text-N", props: {content: "Hello"}}
        - Layout shortcuts: type="row" -> type="container" + layout="horizontal"
        """
        # Simple string becomes text component
        if isinstance(comp, str):
            comp_id = f"text-{self._id_counter}"
            self._id_counter += 1
            return {
                "type": "text",
                "id": comp_id,
                "props": {"content": comp}
            }
        
        # Component object
        if isinstance(comp, dict):
            # Check format: explicit (has "type") or compact (first key is "type#id")
            if "type" in comp:
                # Explicit format - optimal for streaming
                return self._expand_explicit_component(comp)
            
            # Compact format - for hand-written .bp files
            for key, props in comp.items():
                if not isinstance(props, dict):
                    props = {}
                
                # Parse "type#id" or just "type"
                parts = key.split("#", 1)
                comp_type = parts[0]
                comp_id = parts[1] if len(parts) > 1 else f"{parts[0]}-{self._id_counter}"
                if comp_id == parts[0]:
                    self._id_counter += 1
                
                # Convert to explicit format by extracting @ events and children
                explicit_props = {}
                events = {}
                children_data = None
                
                for k, v in props.items():
                    if k.startswith("@"):
                        events[k[1:]] = v
                    elif k == "children":
                        children_data = v
                    else:
                        explicit_props[k] = v
                
                # Build explicit component and recursively expand
                explicit_comp = {
                    "type": comp_type,
                    "id": comp_id,
                    "props": explicit_props
                }
                if events:
                    explicit_comp["on_event"] = events
                if children_data:
                    explicit_comp["children"] = children_data
                
                return self._expand_explicit_component(explicit_comp)
        
        return None
    
    def _expand_explicit_component(self, comp: Dict[str, Any]) -> Dict[str, Any]:
        """
        Expand explicit format component {type, id, props, on_event, children}
        This format is optimal for streaming as it's easy to parse incrementally
        """
        comp_type = comp.get("type", "container")
        comp_id = comp.get("id")
        if not comp_id:
            comp_id = f"{comp_type}-{self._id_counter}"
            self._id_counter += 1
        
        props = comp.get("props", {})
        events = comp.get("on_event", {})
        children_data = comp.get("children", [])
        
        # Handle layout shortcuts on type
        if comp_type == "row":
            comp_type = "container"
            props["layout"] = props.get("layout", "horizontal")
        elif comp_type == "col":
            comp_type = "container"
            props["layout"] = props.get("layout", "vertical")
        elif comp_type in ("sidebar", "main", "editor", "header", "footer", "content", "section"):
            props["role"] = comp_type
            comp_type = "container"
            props["layout"] = props.get("layout", "vertical")
        
        # Recursively expand children
        children = None
        if children_data:
            children = self._expand_components(children_data)
        
        # Build result
        result = {
            "type": comp_type,
            "id": comp_id,
            "props": props
        }
        
        if events:
            result["on_event"] = events
        
        if children:
            result["children"] = children
        
        return result


def parse_blueprint(content: str) -> Dict[str, Any]:
    """
    Convenience function to parse Blueprint content
    
    Args:
        content: Blueprint JSON string
        
    Returns:
        Package dictionary
    """
    parser = BlueprintParser()
    return parser.parse(content)

File: backend/internal/domain/blueprint/parser.go
package blueprint

import (
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/GriffinCanCode/AgentOS/backend/internal/shared/types"
)

// Blueprint represents the root structure of a .bp file
type Blueprint struct {
	App       AppMetadata            `json:"app"`
	Services  []interface{}          `json:"services"`
	UI        map[string]interface{} `json:"ui"`
	Templates map[string]interface{} `json:"templates,omitempty"`
	Config    map[string]interface{} `json:"config,omitempty"`
}

// AppMetadata contains app identification and metadata
type AppMetadata struct {
	ID          string   `json:"id"`
	Name        string   `json:"name"`
	Description string   `json:"description,omitempty"`
	Icon        string   `json:"icon,omitempty"`
	Category    string   `json:"category,omitempty"`
	Version     string   `json:"version"`
	Author      string   `json:"author"`
	Tags        []string `json:"tags,omitempty"`
	Permissions []string `json:"permissions"`
}

// Parser handles Blueprint to Package conversion
type Parser struct {
	templates map[string]interface{}
	idCounter int
}

// NewParser creates a new Blueprint parser
func NewParser() *Parser {
	return &Parser{
		templates: make(map[string]interface{}),
	}
}

// Parse converts Blueprint JSON content to a Package
func (p *Parser) Parse(content []byte) (*types.Package, error) {
	// Reset ID counter for each parse
	p.idCounter = 0

	var bp Blueprint
	if err := json.Unmarshal(content, &bp); err != nil {
		return nil, fmt.Errorf("failed to parse JSON: %w", err)
	}

	// Validate required fields
	if bp.App.ID == "" {
		return nil, fmt.Errorf("app.id is required")
	}
	if bp.App.Name == "" {
		return nil, fmt.Errorf("app.name is required")
	}

	// Default timestamps
	now := time.Now()

	// Store templates for component expansion
	if templates, ok := bp.UI["templates"].(map[string]interface{}); ok {
		p.templates = templates
	}

	// Expand services and extract tool restrictions
	services, restrictions := p.expandServices(bp.Services)

	// Expand UI components
	blueprint := p.expandUISpec(bp.UI)

	return &types.Package{
		ID:                  bp.App.ID,
		Name:                bp.App.Name,
		Description:         bp.App.Description,
		Icon:                bp.App.Icon,
		Category:            bp.App.Category,
		Version:             bp.App.Version,
		Author:              bp.App.Author,
		CreatedAt:           now,
		UpdatedAt:           now,
		Services:            services,
		Permissions:         bp.App.Permissions,
		Tags:                bp.App.Tags,
		ServiceRestrictions: restrictions,
		Blueprint:           blueprint,
	}, nil
}

// expandServices converts service definitions to string array
// Supports: "storage", {storage: [get, set]}, {storage: "*"}
// Returns both the service list and tool restrictions
func (p *Parser) expandServices(services []interface{}) ([]string, []types.ServiceToolRestriction) {
	serviceList := make([]string, 0, len(services))
	restrictions := make([]types.ServiceToolRestriction, 0)

	for _, svc := range services {
		switch v := svc.(type) {
		case string:
			// Simple string: "storage" (all tools allowed)
			serviceList = append(serviceList, v)
			restrictions = append(restrictions, types.ServiceToolRestriction{
				Service:  v,
				AllowAll: true,
			})
		case map[string]interface{}:
			// Object syntax: {storage: [get, set]} or {storage: "*"}
			for key, value := range v {
				serviceList = append(serviceList, key)

				restriction := types.ServiceToolRestriction{
					Service: key,
				}

				switch tools := value.(type) {
				case string:
					if tools == "*" {
						// All tools allowed
						restriction.AllowAll = true
					}
				case []interface{}:
					// Specific tool list
					toolNames := make([]string, 0, len(tools))
					for _, tool := range tools {
						if toolStr, ok := tool.(string); ok {
							toolNames = append(toolNames, toolStr)
						}
					}
					if len(toolNames) > 0 {
						restriction.AllowedTools = toolNames
						restriction.AllowAll = false
					}
				}

				restrictions = append(restrictions, restriction)
			}
		}
	}

	return serviceList, restrictions
}

// expandUISpec converts Blueprint UI to standard UISpec format
func (p *Parser) expandUISpec(ui map[string]interface{}) map[string]interface{} {
	title, _ := ui["title"].(string)
	layout, _ := ui["layout"].(string)
	if layout == "" {
		layout = "vertical"
	}

	lifecycle, _ := ui["lifecycle"].(map[string]interface{})
	components, _ := ui["components"].([]interface{})

	return map[string]interface{}{
		"type":            "app",
		"title":           title,
		"layout":          layout,
		"lifecycle_hooks": p.expandLifecycle(lifecycle),
		"components":      p.expandComponents(components),
	}
}

// expandLifecycle converts lifecycle hooks
func (p *Parser) expandLifecycle(lifecycle map[string]interface{}) map[string]interface{} {
	if lifecycle == nil {
		return map[string]interface{}{}
	}

	result := make(map[string]interface{})
	for hook, action := range lifecycle {
		// Convert hook names (on_mount -> on_mount)
		switch v := action.(type) {
		case string:
			// Single action
			result[hook] = []string{v}
		case []interface{}:
			// Multiple actions
			actions := make([]string, 0, len(v))
			for _, a := range v {
				if str, ok := a.(string); ok {
					actions = append(actions, str)
				}
			}
			result[hook] = actions
		}
	}

	return result
}

// expandComponents recursively expands Blueprint components
func (p *Parser) expandComponents(components []interface{}) []interface{} {
	result := make([]interface{}, 0, len(components))

	for _, comp := range components {
		expanded := p.expandComponent(comp)
		if expanded != nil {
			result = append(result, expanded)
		}
	}

	return result
}

// expandComponent expands a single component with all shortcuts
func (p *Parser) expandComponent(comp interface{}) map[string]interface{} {
	switch v := comp.(type) {
	case string:
		// Simple string: "Hello" -> text component
		compID := fmt.Sprintf("text-%d", p.idCounter)
		p.idCounter++
		return map[string]interface{}{
			"type": "text",
			"id":   compID,
			"props": map[string]interface{}{
				"content": v,
			},
		}

	case map[string]interface{}:
		// Component object
		// Check if this is explicit format (has "type" key) or shorthand format
		if explicitType, hasType := v["type"].(string); hasType {
			// Explicit format: {"type": "button", "id": "...", "props": {...}, "on_event": {...}}
			compType := explicitType
			compID, _ := v["id"].(string)
			propsMap, _ := v["props"].(map[string]interface{})
			if propsMap == nil {
				propsMap = make(map[string]interface{})
			}
			onEventMap, _ := v["on_event"].(map[string]interface{})
			childrenList, _ := v["children"].([]interface{})

			// Build result with explicit format
			result := map[string]interface{}{
				"type":  compType,
				"props": propsMap,
			}

			// Ensure every component has an ID
			if compID != "" {
				result["id"] = compID
			} else {
				result["id"] = fmt.Sprintf("%s-%d", compType, p.idCounter)
				p.idCounter++
			}

			// Add event handlers if present
			if len(onEventMap) > 0 {
				result["on_event"] = onEventMap
			}

			// Recursively expand children if present
			if len(childrenList) > 0 {
				result["children"] = p.expandComponents(childrenList)
			}

			return result
		}

		// Shorthand format: {"button#id": {...props}}
		// Extract type and ID from the first key
		for key, props := range v {
			propsMap, ok := props.(map[string]interface{})
			if !ok {
				continue
			}

			// Parse "type#id" or just "type"
			parts := strings.Split(key, "#")
			compType := parts[0]
			var compID string
			if len(parts) > 1 {
				compID = parts[1]
			}

			// Handle layout shortcuts
			switch compType {
			case "row":
				compType = "container"
				propsMap["layout"] = "horizontal"
			case "col":
				compType = "container"
				propsMap["layout"] = "vertical"
			case "sidebar", "main", "editor", "header", "footer", "content", "section":
				// Semantic container shortcuts - preserve semantic meaning via role
				role := compType
				compType = "container"
				propsMap["role"] = role // Add role for styling/identification
				if _, hasLayout := propsMap["layout"]; !hasLayout {
					propsMap["layout"] = "vertical"
				}
			}

			// Check for template reference
			if templateName, ok := propsMap["$template"].(string); ok {
				if template, exists := p.templates[templateName]; exists {
					if templateMap, ok := template.(map[string]interface{}); ok {
						// Merge template with current props (current overrides template)
						merged := make(map[string]interface{})
						for k, v := range templateMap {
							merged[k] = v
						}
						for k, v := range propsMap {
							if k != "$template" {
								merged[k] = v
							}
						}
						propsMap = merged
					}
				}
			}

			// Extract event handlers, children, and directives
			events := make(map[string]interface{})
			cleanProps := make(map[string]interface{})
			var children []interface{}
			var conditional interface{}
			var loopConfig interface{}

			for k, v := range propsMap {
				if strings.HasPrefix(k, "@") {
					// Event handler: @click -> click
					eventName := strings.TrimPrefix(k, "@")
					events[eventName] = v
				} else if k == "children" {
					// Recursively expand children
					if childList, ok := v.([]interface{}); ok {
						children = p.expandComponents(childList)
					}
				} else if k == "$if" {
					conditional = v
				} else if k == "$for" {
					loopConfig = v
				} else if !strings.HasPrefix(k, "$") {
					// Regular props (skip other $ directives)
					cleanProps[k] = v
				}
			}

			// Build result
			result := map[string]interface{}{
				"type":  compType,
				"props": cleanProps,
			}

			// Ensure every component has an ID (required by frontend)
			if compID != "" {
				result["id"] = compID
			} else {
				// Generate auto ID: type-N (e.g., "container-0", "button-1")
				result["id"] = fmt.Sprintf("%s-%d", compType, p.idCounter)
				p.idCounter++
			}

			if len(events) > 0 {
				result["on_event"] = events
			}

			if len(children) > 0 {
				result["children"] = children
			}

			// Add conditional/loop metadata (frontend will handle)
			if conditional != nil {
				result["$if"] = conditional
			}

			if loopConfig != nil {
				result["$for"] = loopConfig
			}

			return result
		}
	}

	return nil
}

// ParseFile is a convenience function for parsing Blueprint files
func ParseFile(content []byte) (*types.Package, error) {
	parser := NewParser()
	return parser.Parse(content)
}
